### ISSUE TITLE
Constructor calls (including with parameters) are translated as long as there are initializations

### ISSUE THREAD

**Issue by roji** [dotnet/efcore#31315]
For #31279, I looked at our translation policy around constructors:

```c#
// FAILS (GOOD)
_ = ctx.Blogs
    .Select(b => new Blog(1))
    .Where(b => b.Name == "sdf")
    .ToList();

// WORKS (GOOD)
_ = ctx.Blogs
    .Select(b => new Blog { X = Y })
    .Where(b => b.Name == "sdf")
    .ToList();

// FAILS (OK, but somewhat inconsistent with the above - could just translate to an empty Blog etc.)
_ = ctx.Blogs
    .Select(b => new Blog())
    .Where(b => b.Name == "sdf")
    .ToList();

// WORKS (PROBLEMATIC)
_ = ctx.Blogs
    .Select(b => new Blog(1) { X = Y })
    .Where(b => b.Name == "sdf")
    .ToList();
```

It seems problematic to allow constructors with parameters when initializers are present. We obviously can't do anything with the arguments so they are silently ignored, although users may mistakenly think that they're being assigned to some property.

As a secondary thing, it's a bit odd to disallow `new Blog()` when we allow `new Blog { X = Y }`; after all in the latter case there may be other properties on Blog which are left uninitialized - so why not allow that for all properties with `new Blog()`. But it's probably not very useful as a query feature.

**Comment by ajcvickers** [comment]
> We obviously can't do anything with the arguments so they are silently ignored, 

Why can't we bind to the parameters?

**Comment by roji** [comment]
I think the main issue would be the assumption that constructor is simply pass-through, i.e. assigning the argument to the corresponding property (there could be logic in there that e.g. transforms it).

**Comment by ajcvickers** [comment]
@roji Yeah, but we call parameterized constructors when creating entity instances not in projections.

**Comment by ajcvickers** [comment]
I must have misunderstood what you were saying. Given this entity type:

```C#
public class Blog
{
    public Blog(int id, string? name)
    {
        Id = id;
        Name = name;
    }

    public int Id { get; set; }
    public string? Name { get; set; }
}
```

I thought you were saying we didn't support this:

```C#
var blog2 = context.Blogs.Select(e => new Blog(e.Id, e.Name)).Single();
```

And I was wondering why not given that we support:

```C#
var blog1 = context.Blogs.Single();
```

But I just checked and they both work.

**Comment by roji** [comment]
Yeah. That Select works only because of client evaluation (no operator after), so the constructor really does execute in .NET.

**Comment by ajcvickers** [comment]
Ah, okay. So, then my question stands: why could we not make both these things work?

```C#
var blog1 = context.Blogs.OrderBy(e => e.Id).Single();
var blog2 = context.Blogs.Select(e => new Blog(e.Id, e.Name)).OrderBy(e => e.Id).Single();
```

**Comment by roji** [comment]
Because of this, no?


> I think the main issue would be the assumption that constructor is simply pass-through, i.e. assigning the argument to the corresponding property (there could be logic in there that e.g. transforms it).

Since the Select isn't last, we need to translate the lambda rather than client-evaluate it; at that point we'd be assuming that the constructor has no special logic (where we don't need to make that assumption with client-eval). Or are you saying that this assumption seems reasonable to you?

**Comment by ajcvickers** [comment]
I'm saying that we already make that assumption when we do call the constructor--like in the first query above. That is, we bound `id` to `Id` and `name` to `Name`, but we have no way of knowing what _really happens_ to these values when the constructor is called.

This used to bother me. But, so far, no body has ever complained that we got it wrong.

**Comment by roji** [comment]
> I'm saying that we already make that assumption when we do call the constructor--like in the first query above.

I may be misunderstanding, but are you talking about this one: `context.Blogs.OrderBy(e => e.Id).Single()`? If so, here the constructor is called as part of materialization, so it's evaluated in .NET. Whereas in the 2nd query, if the OrderBy is to be translated, we'd have to also translate the constructor beforehand?

We're probably talking over each other's heads...

**Comment by roji** [comment]
Let's discuss in design to be sure we're understanding each other...

**Comment by ajcvickers** [comment]
Note from triage: it is reasonable to translate using calls to a parameterized constructor as part of ensuring all mapped properties have been initialized. If this is done internally in the query, then it is the same as using the properties directly and will be translated as such--that is, assuming no side effects.
